### PR TITLE
[client-go] Generate from outside GOPATH

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Make client-go generate work from outside GOPATH and a reference to k8s.io/code-generator in go.mod

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: hack/internal/tools/pinversion.go
Comment (tenzen-y) [inline hack/internal/tools/pinversion.go:1]: I remember that we decided not to use tools after discussing it with @alculquicondor.

cc: @alculquicondor

**Comment by tenzen-y** [comment]
NOTE: Even if using the main branch, I could generate the client-go code if I put the kueue repository the under GOPATH.

File: hack/update-codegen.sh [hunk 1]
```diff
@@ -18,16 +18,28 @@ set -o errexit
...
+# for now we should just move the generated code to `client-go`.
+if  [ -d sigs.k8s.io/kueue/client-go ]
+then
+  echo "Out of GOPATH generation, move the generated files to ../client-go"
+  if  [ -d ${KUEUE_ROOT}/client-go ]
+  then
+    rm -r ${KUEUE_ROOT}/client-go
+  fi
+  mv sigs.k8s.io/kueue/client-go ${KUEUE_ROOT}/client-go
+fi
```
Comment (alculquicondor) [inline hack/update-codegen.sh:45]: why is this necessary now but it wasn't before?

File: hack/internal/tools/pinversion.go
Comment (alculquicondor) [reply hack/internal/tools/pinversion.go:1]: I guess in this case both options are not pretty.

We can't use `go install` because it wouldn't include the `.sh` file.

Using the `tools` package is not consistent with the rest of the tools.

Is there one approach that would allow us for every tool to be installed the same way?

File: hack/update-codegen.sh (hunk 1 above)
Comment (alculquicondor) [reply hack/update-codegen.sh:45]: You mean GOPATH?

**Comment by tenzen-y** [comment]
> The scope is not to move the code in GOROOT.

I meant I could generate codes in GOPATH without this improvement.

**Comment by trasc** [comment]
> > The scope is not to move the code in GOROOT.
> 
> I meant I could generate codes in GOPATH without this improvement.
> 

I know, I need it to work outside GOPATH.

**Comment by tenzen-y** [comment]
> > > The scope is not to move the code in GOROOT.
> > 
> > 
> > I meant I could generate codes in GOPATH without this improvement.
> 
> I know, I need it to work outside GOPATH.

I'm a bit confused. Which do you mean GOROOT or GOPATH?

**Comment by trasc** [comment]
GOPATH, sorry for the confusion. (It's been some time since I needed to use this term (before  go 1.11))

**Comment by alculquicondor** [comment]
I'm ok with this change, if we decide that we will move all tools to use the same approach as this PR, instead of `go install`

**Comment by tenzen-y** [comment]
> I'm ok with this change, if we decide that we will move all tools to use the same approach as this PR, instead of `go install`

+1
Because we should use consistent ways to manage tools.

**Comment by trasc** [comment]
The go install step cannot be removed, but we can have the same way of managing the versions.(via go.mod)

**Comment by alculquicondor** [comment]
Leaving LGTM to @tenzen-y

File: hack/update-codegen.sh [hunk 2]
```diff
@@ -18,16 +18,28 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
-cd "$(dirname "${0}")"
 GO_CMD=${1:-go}
-CODEGEN_PKG=${2:-../bin}
+KUEUE_ROOT=$(realpath $(dirname ${BASH_SOURCE[0]})/..)
+CODEGEN_PKG=$($GO_CMD list -m -f "{{.Dir}}" k8s.io/code-generator )
```
Comment (tenzen-y) [inline hack/update-codegen.sh:23]:
```suggestion
CODEGEN_PKG=$($GO_CMD list -m -f "{{.Dir}}" k8s.io/code-generator)
```

File: hack/update-codegen.sh [hunk 3]
```diff
@@ -18,16 +18,28 @@ set -o errexit
...
-bash "${CODEGEN_PKG}/generate-groups.sh" \
+cd $(dirname ${BASH_SOURCE[0]})/..
+
+source "${CODEGEN_PKG}/generate-groups.sh" \
   "all" \
   sigs.k8s.io/kueue/client-go \
   sigs.k8s.io/kueue/apis \
   kueue:v1beta1 \
-  --go-header-file ./boilerplate.go.txt
+  --go-header-file ${KUEUE_ROOT}/hack/boilerplate.go.txt \
```
Comment (tenzen-y) [inline hack/update-codegen.sh:32]:
```suggestion
  --go-header-file ${KUEUE_ROOT}/hack/boilerplate.go.txt
```

**Review by tenzen-y — COMMENTED** [review]
I left comments for nits.
